flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/adaptive/AdaptiveJoinOperatorGenerator.java

```diff
@@ -135,50 +148,60 @@ public StreamOperatorFactory<?> genOperatorFactory(
+            Long threshold,
+            Function<Boolean, Boolean> tryTransformEdges) {
         boolean leftSizeSmallerThanThreshold = leftInputSize <= threshold;
         boolean rightSizeSmallerThanThreshold = rightInputSize <= threshold;
         boolean leftSmallerThanRight = leftInputSize < rightInputSize;
+        boolean isBroadcastJoinTemp;
+        boolean leftIsBuildTemp;
         switch (joinType) {
             case RIGHT:
                 // For a right outer join, if the left side can be broadcast, then the left side is
                 // always the build side; otherwise, the smaller side is the build side.
-                isBroadcastAndLeftBuild =
-                        new Tuple2<>(
-                                leftSizeSmallerThanThreshold,
-                                leftSizeSmallerThanThreshold ? true : leftSmallerThanRight);
+                isBroadcastJoinTemp = leftSizeSmallerThanThreshold;
+                leftIsBuildTemp = isBroadcastJoinTemp ? true : leftSmallerThanRight;
                 break;
             case INNER:
-                isBroadcastAndLeftBuild =
-                        new Tuple2<>(
-                                leftSizeSmallerThanThreshold || rightSizeSmallerThanThreshold,
-                                leftSmallerThanRight);
+                isBroadcastJoinTemp = leftSizeSmallerThanThreshold || rightSizeSmallerThanThreshold;
+                leftIsBuildTemp = leftSmallerThanRight;
                 break;
             case LEFT:
             case SEMI:
             case ANTI:
                 // For left outer / semi / anti join, if the right side can be broadcast, then the
                 // right side is always the build side; otherwise, the smaller side is the build
                 // side.
-                isBroadcastAndLeftBuild =
-                        new Tuple2<>(
-                                rightSizeSmallerThanThreshold,
-                                rightSizeSmallerThanThreshold ? false : leftSmallerThanRight);
+                isBroadcastJoinTemp = rightSizeSmallerThanThreshold;
+                leftIsBuildTemp = isBroadcastJoinTemp ? false : leftSmallerThanRight;
                 break;
             case FULL:
             default:
                 throw new RuntimeException(String.format("Unexpected join type %s.", joinType));
         }
 
-        isBroadcastJoin = isBroadcastAndLeftBuild.f0;
-        leftIsBuild = isBroadcastAndLeftBuild.f1;
+        if (isBroadcastJoinTemp && tryTransformEdges.apply(leftIsBuildTemp)) {
+            isBroadcastJoin = true;
+            leftIsBuild = leftIsBuildTemp;
+        } else {
+            isBroadcastJoin = false;
+            leftIsBuild = leftSmallerThanRight;
+        }
+
+        return new Tuple2<>(isBroadcastJoin, leftIsBuild);
+    }
+
+    @Override
+    public boolean isLeftBuild() {
         // Sort merge join requires the left side to be read first if the broadcast threshold is not
         // met.
```

Comment:
-> Sort merge join requires the left side to be read first. Therefore, if it cannot be converted into a BroadcastHashJoin, we should mark its left side as the build side.